### PR TITLE
Feature/2037 user input settings datastore

### PR DESCRIPTION
**Pull request by eugene-manuilov**
## Summary

Addresses issue #2037

## Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

### CODE REVIEW

File: assets/js/googlesitekit/datastore/user/index.js [hunk 1]
```diff
@@ -23,6 +23,7 @@ import Data from 'googlesitekit-data';
 import authentication from './authentication';
 import dateRange from './date-range';
 import disconnect from './disconnect';
+import inputSettings from './input-settings';
```
Comment (felixarntz) [inline assets/js/googlesitekit/datastore/user/index.js:26]: Nitpicking, but let's call this file `user-input-settings.js` since "User Input" is kind of a semi-internal codename for this feature. "input" on its own therefore isn't as clear here.

File: assets/js/googlesitekit/datastore/user/input-settings.js [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+}
+
+const fetchGetUserInputSettingsStore = createFetchStore( {
+	baseName: 'getUserInputSettings',
+	controlCallback: () => API.get( 'core', 'user', 'user-input-settings', undefined, { useCache: false } ),
+	reducerCallback: fetchStoreReducerCallback,
+} );
+
+const fetchSaveUserInputSettingsStore = createFetchStore( {
+	baseName: 'setUserInputSettings',
```
Comment (felixarntz) [inline assets/js/googlesitekit/datastore/user/input-settings.js:50]: This should be consistently named, i.e. `saveUserInputSettings`.

File: assets/js/googlesitekit/datastore/user/input-settings.js [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param {Object} settings User input settings.
+	 * @return {Object} Object with `response` and `error`.
+	 */
+	*saveUserInputSettings( settings ) {
+		invariant( isPlainObject( settings ), 'a valid settings are required.' );
+
+		const { response, error } = yield fetchSaveUserInputSettingsStore.actions.fetchSetUserInputSettings( settings );
```
Comment (felixarntz) [inline assets/js/googlesitekit/datastore/user/input-settings.js:113]: See above, references need to be adjusted accordingly.

File: assets/js/googlesitekit/datastore/user/input-settings.test.js [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+					data: { status: 500 },
+				};
+
+				fetchMock.post(
+					coreUserInputSettingsEndpointRegExp,
+					{ body: response, status: 500 }
+				);
+
+				await registry.dispatch( STORE_NAME ).saveUserInputSettings( ...args );
+				expect( registry.select( STORE_NAME ).getErrorForAction( 'setUserInputSettings', args ) ).toMatchObject( response );
```
Comment (felixarntz) [inline assets/js/googlesitekit/datastore/user/input-settings.test.js:144]: See above.

**Review by felixarntz — COMMENTED** [review]
Great! Just some inconsistent naming here, I'll quickly update that.

File: assets/js/googlesitekit/datastore/user/user-input-settings.js [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+
+	/**
+	 * Saves user input settings.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param {Object} settings User input settings.
+	 * @return {Object} Object with `response` and `error`.
+	 */
+	*saveUserInputSettings( settings ) {
```
Comment (felixarntz) [inline assets/js/googlesitekit/datastore/user/user-input-settings.js:110]: Per the ACs, this action should be parameter-less. It should simply save the current settings from the client. In other words, the functionality should match `saveSettings` from `createSettingsStore` - including the need for manually doing a `receiveError` call because this action doesn't require arguments while the internal action does.

**Review by felixarntz — CHANGES_REQUESTED** [review]
There's one thing that needs to be slightly reworked here.

File: assets/js/googlesitekit/datastore/user/user-input-settings.js (hunk 5 above)
Comment (eugene-manuilov) [reply assets/js/googlesitekit/datastore/user/user-input-settings.js]: Updated.

File: assets/js/googlesitekit/datastore/user/user-input-settings.js [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+	 * @return {Object} Object with `response` and `error`.
+	 */
+	*saveUserInputSettings() {
+		const registry = yield Data.commonActions.getRegistry();
+		const settings = registry.select( STORE_NAME ).getUserInputSettings();
+
+		const { response, error } = yield fetchSaveUserInputSettingsStore.actions.fetchSaveUserInputSettings( settings );
+		if ( error ) {
+			// Store error manually since saveUserInputSettings signature differs from fetchSaveUserInputSettings.
+			registry.dispatch( STORE_NAME ).receiveError( error, 'saveUserInputSettings' );
```
Comment (felixarntz) [inline assets/js/googlesitekit/datastore/user/user-input-settings.js:116]: This is somewhat incorrect and would actually need to be like this:
```suggestion
			registry.dispatch( STORE_NAME ).receiveError( error, 'saveUserInputSettings', [] );
```

However, this is an unrelated problem and will be fixed via #2210, so we can ignore it here.

**Review by felixarntz — APPROVED** [review]
